### PR TITLE
CMake: Multi-Config Generator

### LINKED ISSUE #1383: Examples: Fix Windows PDB
With the `dev` branch pre 0.15.0:
```
C:\Users\runneradmin\AppData\Local\Temp\ci-pU5qdA784w\src\examples\10_streaming_write.cpp : fatal  error C1041: cannot open program database 'C:\Users\runneradmin\AppData\Local\Temp\ci-pU5qdA784w\build\bin\RelWithDebInfo\vc143.pdb'; if multiple CL.EXE write to the same .PDB file, please use /FS [C:\Users\runneradmin\AppData\Local\Temp\ci-pU5qdA784w\build\10_streaming_write.vcxproj]
  Building Custom Rule C:/Users/runneradmin/AppData/Local/Temp/ci-pU5qdA784w/src/CMakeLists.txt
  Building Custom Rule C:/Users/runneradmin/AppData/Local/Temp/ci-pU5qdA784w/src/CMakeLists.txt
  12_span_write.cpp
C:\Users\runneradmin\AppData\Local\Temp\ci-pU5qdA784w\src\examples\12_span_write.cpp : fatal  error C1041: cannot open program database 'C:\Users\runneradmin\AppData\Local\Temp\ci-pU5qdA784w\build\bin\RelWithDebInfo\vc143.pdb'; if multiple CL.EXE write to the same .PDB file, please use /FS [C:\Users\runneradmin\AppData\Local\Temp\ci-pU5qdA784w\build\12_span_write.vcxproj]
...
```

First seen in https://github.com/ax3l/cmake-easyinstall/pull/12

Regression to #1326

### PR DESCRIPTION
**Pull request by ax3l**
Set target properties for multi-config generators.

Fix #1383

Regression to #1326

- [x] test in https://github.com/ax3l/cmake-easyinstall/pull/12

### CODE REVIEW

**Comment by ax3l** [comment]
Hm, does not yet work. Seems to actually collide on a `vc143.pdb` file in the runtime dir...

Update: solved by setting the `COMPILE_PDB_NAME[_CFG]` output name target properties.
Contrary to `PDB_NAME`, the default here is unspecified and seems to default to `vc143.pdb`

File: CMakeLists.txt [hunk 1]
```diff
@@ -573,6 +575,19 @@ set_target_properties(openPMD PROPERTIES
...
+# note: same as above, but for Multi-Config generators
+if(isMultiConfig)
+    foreach(CFG IN LISTS CMAKE_CONFIGURATION_TYPES)
+        string(TOUPPER "${CFG}" CFG_UPPER)
+        set_target_properties(openPMD PROPERTIES
+            ARCHIVE_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_ARCHIVE_OUTPUT_DIRECTORY}
+            LIBRARY_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_LIBRARY_OUTPUT_DIRECTORY}
+            RUNTIME_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_RUNTIME_OUTPUT_DIRECTORY}
+            PDB_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_PDB_OUTPUT_DIRECTORY}
+            COMPILE_PDB_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_COMPILE_PDB_OUTPUT_DIRECTORY}
```
Comment (franzpoeschel) [inline CMakeLists.txt]:
```suggestion
            COMPILE_PDB_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_COMPILE_PDB_OUTPUT_DIRECTORY}/${CFG}
```
Weren't the build paths on Windows something like this?

**Comment by ax3l** [comment]
@franzpoeschel ok, now the parallel build works and is not a hack anymore :)

**Comment by franzpoeschel** [comment]
Nice!

So, what was missing is the `COMPILE_PDB_NAME` so that the different processes don't try accessing the same file?

**Comment by ax3l** [comment]
Yes, exactly. Seems it was unspecified and collided for different executables we build in parallel.
I added an update here: https://github.com/openPMD/openPMD-api/pull/1384#issuecomment-1450968569

**Comment by franzpoeschel** [comment]
I actually saw a similar issue in another of my PRs today (the eternal #824), I'll have to rebase and see if this fixes it :)

**Review by franzpoeschel — APPROVED** [review]
Thanks!

File: CMakeLists.txt [hunk 2]
```diff
@@ -835,12 +882,27 @@ if(openPMD_HAVE_PYTHON)
...
         COMPILE_PDB_OUTPUT_DIRECTORY ${openPMD_PYTHON_OUTPUT_DIRECTORY}/openpmd_api
     )
+    # note: same as above, but for Multi-Config generators
+    if(isMultiConfig)
+        foreach(CFG IN LISTS CMAKE_CONFIGURATION_TYPES)
+            string(TOUPPER "${CFG}" CFG_UPPER)
+            set_target_properties(openPMD.py PROPERTIES
+                COMPILE_PDB_NAME_${CFG_UPPER} openpmd_api_cxx
+                ARCHIVE_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_PYTHON_OUTPUT_DIRECTORY}/${CFG}/openpmd_api
+                LIBRARY_OUTPUT_DIRECTORY_${CFG_UPPER} ${openPMD_PYTHON_OUTPUT_DIRECTORY}/${CFG}/openpmd_api
```
Comment (ax3l) [inline CMakeLists.txt:899]: This broke my slightly arcane logic in `setup.py` for Windows.
Fix in #1400

**Pull request merged**